core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -396,7 +396,26 @@ abstract class RDD[T: ClassTag](
    * Return a new RDD containing the distinct elements in this RDD.
    */
   def distinct(numPartitions: Int)(implicit ord: Ordering[T] = null): RDD[T] = withScope {
-    map(x => (x, null)).reduceByKey((x, y) => x, numPartitions).map(_._1)
+    partitioner match {
+      case Some(p) if numPartitions == partitions.length =>
+        def key(x: T): (T, Null) = (x, null)
+        val cleanKey = sc.clean(key _)
+        val keyed = new MapPartitionsRDD[(T, Null), T](
+          this,
+          (context, pid, iter) => iter.map(cleanKey),
+          knownPartitioner = Some(new WrappedPartitioner(p)))
+        val duplicatesRemoved = keyed.reduceByKey((x, y) => x)
```

Comment:
Dont you need to specify the partitioner here ?

Reply:
No reduceByKey on a known partioner is fine.

Reply:
Ah yes, no partitioner specified => use parent's partitioner.

Reply:
So I _think_ it is partitioner of input RDD if known partitioner otherwise hash partitioner of the default parallelism. Yes?

Reply:
Yes, would use right partitioner in this case